### ISSUE TITLE
Extraneous control flow error

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#37480]
From vscode: src/vs/workbench/contrib/search/common/searchModel.ts

**Code**

```ts
interface F {
    d(): void
}
export function doRemove(dds: F | F[]) {
    if (!Array.isArray(dds)) {
        dds = [dds]
    }
    for (let n of dds) {
        n.d()
    }
    return dds
}
```

**Expected behavior:**

No error, and `n: F` at all locations, and `dds: F[]` at all locations.

**Actual behavior:**

Types are correct; bogus error: 'F | F[]' must have a '[Symbol.iterator]()' method that returns an iterator.

Looks like narrowing works as expected, except for errors.

**Comment by sandersn** [comment]
For future bisecting: Doesn't repro on a commit from March 2: dd6811fbe37f05a2b79427112a919d75a7417176
